ufl/core/base_form_operator.py

```diff
@@ -182,4 +183,31 @@ def __hash__(self):
 
     def __eq__(self, other):
         """Check for equality."""
-        raise NotImplementedError()
+        if self is other:
+            return True
+        return (
+            type(self) is type(other)
+            and all(a == b for a, b in zip(self._argument_slots, other._argument_slots))
+            and self.ufl_function_space() == other.ufl_function_space()
+        )
+
+    @property
+    def _parent_type(self):
+        """Is this a primal or dual expression?"""
+        return BaseForm if is_dual(self.ufl_function_space()) else Operator
+
+    def __add__(self, other):
+        """Add."""
+        return self._parent_type.__add__(self, other)
```

Comment:
Is this the right way of dealing with multiple inheritance dynamically?

Reply:
I think this is an OK thing to do.